### PR TITLE
MANTA-5368 rebalancer should be able to read directly from a cloned DB

### PR DESCRIPTION
**Pull request by rjloura**
This change allows sharkspotter query objects directly from a postgres database.  The `config::Config` field `direct_db` or the command line flag `-D` will enable this feature.  For now the postgres database hostnames are hardcoded to `<shard>.rebalancer-postgres.<domain>`.  This is intentional in order to minimize (nearly eliminate) the possibility that sharkspotter will read directly from the manta postgres primary or sync database. This means in order for the feature to work a new, cloned, database will need to be served from that hostname.  The creation of this database can be accomplished with the script under `tools/pgclone.sh`.

Also note that there are some differences in the format of moray manta table records provided by moray and the postgres database schema.  Because of this a new `MorayMantaBucketObject` structure was added to deserialize the records returned from the database.   The `SharkspotterMessage` fields were also modified to remove the `value` (moray value) and replace it with `manta_value` and `etag`.  The initial change that brought in the full moray value (MANTA-5249) was intended to provide the full manta object metadata record entry which is found under the moray `_value` column/field as well as the etag.  Since there is divergnce in the moray and database schema formats and since both have a `_value` and an `_etag` entry and since we do not need the full moray entry it was decided to only send back what we need.  This also allows us to do less deserializing (which was found to have a non-negligible performance impact) on both sharkspotter and on the rebalancer. 

Finally note that the query used to gather records is `SELECT * FROM manta WHERE type = 'object'`.  The method used to make the query returns a TCP Stream which is asynchronously read.  For each row we parse the record, determine if it identifies an object on a shark that we are interested in, and then formats a `SharkspotterMessage` which is sent down a crossbeam channel (to the rebalancer).  During testing it was noted that if the database connection is interrupted we will continue scanning the other shards, and report & return Error at the end.

### CODE REVIEW

File: src/directdb.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    shard: u32,
+    conf: Config,
+    log: Logger,
+    obj_tx: crossbeam::Sender<SharkspotterMessage>,
+) -> Result<(), Error> {
+    let shard_host_name =
+        format!("{}.rebalancer-postgres.{}", shard, conf.domain);
+
+    debug!(log, "Connecting to {}", shard_host_name);
+    let (client, connection) = tokio_postgres::Config::new()
```
Comment (rhb2) [inline src/directdb.rs:71]: Had to go through the documentation to figure out what this was doing.  I'm consulting [the tokio_postgres 0.5.5 documentation](https://docs.rs/tokio-postgres/0.5.5/tokio_postgres/) as I go through.  A little bit of commentary around what's happening would be helpful.

File: src/directdb.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        connection
+            .await
+            .map_err(|e| Error::new(ErrorKind::Other, e))?;
+        Ok::<(), Error>(())
+    });
+
+    let rows = client
+        .query_raw("SELECT * from manta where type='object'", vec![])
+        .await
+        .map_err(|e| Error::new(ErrorKind::Other, e))?;
```
Comment (rhb2) [inline src/directdb.rs:90]: Is it worth logging the number of rows in the result on success?

File: src/directdb.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                    shark: s.manta_storage_id.clone(),
+                    shard,
+                };
+
+                if let Err(e) = obj_tx.send(msg) {
+                    warn!(log, "Tx channel disconnected: {}", e);
+                    return Err(Error::new(ErrorKind::BrokenPipe, e));
+                }
+                Ok(())
+            })
```
Comment (rhb2) [inline src/directdb.rs:139]: This is mostly just me trying to understand what's going on rather then critique.

My understanding:
This massive loop is iterating through each row in the results from the query, filtering by a specific storage_id (presumably the one of the shark that we want to evacuate in the context of the rebalancer at least), then it's extracting the necessary information from it in order to populate the `SharkspotterMessage` object, which it then forwards on to the client?

If my understanding is wrong, then disregard the rest.  My opinion, and it's optional, would be to break a good chunk of this loop out in to a function that we can also test by itself by faking up some data, and running it through the function and validating that we get what we expect at the other end.  I'd probably advocate that more strongly if this was going to be around indefinitely, but it's not, so I leave it up to you.

File: src/main.rs [hunk 4]
```diff
@@ -81,7 +69,7 @@ where
         + FnMut(SharkspotterMessage) -> Result<(), Error>,
 {
     let channel: (Sender<SharkspotterMessage>, Receiver<SharkspotterMessage>) =
-        crossbeam_channel::bounded(10);
+        crossbeam_channel::bounded(100);
```
Comment (rhb2) [inline src/main.rs:72]: I don't have any issues with this per say, but just want to know why this got increased this by an order?

**Comment by rjloura** [comment]
> One more general concern that I have, and maybe it's not valid for some reason unknown to me, is there anything preventing this thing from directly interrogating one of the pg databases that we're using in a manatee cluster? I assume that zookeeper will have no knowledge of the cloned pg database that we're using.
> 
> Other than that, my feedback is pretty superficial. If I understand the change correctly and it works, then I'm ready when you are.

That's a legitimate concern.  If you look at `directdb.rs` line 68 you'll see that I hard coded a portion of the host name
```
let shard_host_name = format!("{}.rebalancer-postgres.{}", shard, conf.domain);
```

I thought it might be useful to allow this to be overridden by a config param or CLI option, but for now I'd like to leave it as is.

This hostname should be published in ZK, but no other service should know about it.

File: src/directdb.rs (hunk 3 above)
Comment (rjloura) [reply src/directdb.rs:139]: You're understanding is correct.  I refactored some of this.

File: src/main.rs (hunk 4 above)
Comment (rjloura) [reply src/main.rs:72]: This is currently only used in testing sharkspotter performance.  Considering we have upwards of 100 shards in a given region, I figured a deeper queue would keep the file writing thread a bit more busy and would reduce the amount of time the per-shard scanning threads are stuck blocked on the writing thread.

In the rebalancer PR I increased this queue depth / channel size as well.

File: src/directdb.rs (hunk 1 above)
Comment (rjloura) [reply src/directdb.rs:72]: added some comments.

File: src/directdb.rs (hunk 2 above)
Comment (rjloura) [reply src/directdb.rs:91]: This is a stream.  I don't know if maintaining a counter is that helpful here.  We already do that in rebalancer.

File: Cargo.toml [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
 edition = "2018"
 
 [dependencies]
 assert_cli = "0.6.0"
-serde = { version = "1.0.89", features = ["derive"] }
-serde_json = "1.0.39"
+
+# TODO version numbers
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
```
[the remark this answers is not in the document]
Comment (rjloura) [reply Cargo.toml]: There were some dependency conflicts that I had trouble figuring out so I just removed the minor versions.  I added them back to a more recent version.

File: src/directdb.rs (hunk 2 above)
Comment (rhb2) [reply src/directdb.rs:91]: Oh!  I remember that now.  Subconsciously, I think that's what motivated this comment.

**Pull request merged**